ext/semian_cb_data/semian_cb_data.c

```diff
@@ -0,0 +1,900 @@
+    //rb_warn("semctl: semaphore set with semid %d deleted", ptr->semid);
+    ptr->semid = -1;
+  }
+  return self;
+}
+
+
+
+
+/*
+ * semian_cb_data_lock/unlock and associated functions decrement/increment semaphore
+ */
+
+static VALUE
+semian_cb_data_lock(VALUE self)
+{
+  return (VALUE) WITHOUT_GVL(semian_cb_data_lock_without_gvl, (void *)self, RUBY_UBF_IO, NULL);
+}
+
+static void *
+semian_cb_data_lock_without_gvl(void *self)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct((VALUE)self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (ptr->lock_triggered)
+    return (void *)Qtrue;
+  if (-1 == ptr->semid){
+    rb_raise(eInternal, "semid not set, errno %d: (%s)", errno, strerror(errno));
+    return (void *)Qfalse;
+  }
+  VALUE retval;
+
+  struct timespec ts = { 0 };
+  ts.tv_sec = kCBInternalTimeout;
+
+  if (-1 == semtimedop(ptr->semid,&decrement,1, &ts)) {
+    rb_raise(eInternal, "error with semop locking,  %d: (%s)", errno, strerror(errno));
+    retval=Qfalse;
+  } else
+    retval=Qtrue;
+
+  ptr->lock_triggered = true;
+  //rb_warn("semop: lock success");
+  return (void *)retval;
+}
+
+static VALUE
+semian_cb_data_unlock(VALUE self)
+{
+  return (VALUE) WITHOUT_GVL(semian_cb_data_unlock_without_gvl, (void *)self, RUBY_UBF_IO, NULL);
+}
+
+static void *
+semian_cb_data_unlock_without_gvl(void *self)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct((VALUE)self, semian_cb_data, &semian_cb_data_type, ptr);
+  if (!(ptr->lock_triggered))
+    return (void *)Qtrue;
+  if (-1 == ptr->semid){
+    rb_raise(eInternal, "semid not set, errno %d: (%s)", errno, strerror(errno));
+    return (void *)Qfalse;
+  }
+  VALUE retval;
+
+  struct timespec ts = { 0 };
+  ts.tv_sec = kCBInternalTimeout;
+
+  if (-1 == semtimedop(ptr->semid,&increment,1 , &ts)) {
+    rb_raise(eInternal, "error with semop unlocking, errno: %d (%s)", errno, strerror(errno));
+    retval=Qfalse;
+  } else
+    retval=Qtrue;
+
+  ptr->lock_triggered = false;
+  //rb_warn("semop unlock success");
+  return (void *)retval;
+}
+
+
+
+
+/*
+  Acquire memory by getting shmid, and then attaching it to a memory location,
+    requires semaphore for locking/unlocking to be setup
+*/
+static VALUE
+semian_cb_data_acquire_memory(VALUE self, VALUE permissions)
+{
+  semian_cb_data *ptr;
+  TypedData_Get_Struct(self, semian_cb_data, &semian_cb_data_type, ptr);
+
+  if (-1 == ptr->semid){
+    rb_raise(eInternal, "semid not set, errno %d: (%s)", errno, strerror(errno));
+    return self;
+  }
+  if (TYPE(permissions) != T_FIXNUM)
+    rb_raise(rb_eTypeError, "expected integer for permissions");
+
+  if (!semian_cb_data_lock(self))
+    return self;
+
+  key_t key = ptr->key;
+  if (-1 == (ptr->shmid = shmget( key,
+                                  2*sizeof(int) + ptr->arr_max_size * sizeof(double),
+                                  IPC_CREAT | IPC_EXCL | FIX2LONG(permissions)))) {
+    if (errno == EEXIST)
+      ptr->shmid = shmget(key, ptr->arr_max_size, IPC_CREAT);
```

Comment:
I've always assumed `shmget` needed `IPC_CREAT` to actually fetch the shmid, although it makes sense for it to not need it. But I was also following along with what was similarly done in `create_semaphore()` from `semian.c` which was just to remove the `IPC_EXCL` flag.
```
static int
create_semaphore(int key, int permissions, int *created)
{
  int semid = 0;
  int flags = 0;

  *created = 0;
  flags = IPC_EXCL | IPC_CREAT | permissions;

  semid = semget(key, kCBSemaphoreCount, flags);
  if (semid >= 0) {
    *created = 1;
  } else if (semid == -1 && errno == EEXIST) {
    flags &= ~IPC_EXCL;
    semid = semget(key, kCBSemaphoreCount, flags);
  }
  return semid;
}
```

.. which means the code should really be this. It's safer just to have the `IPC_CREAT` flag.
```
  key_t key = ptr->key;
  int byte_size = 2*sizeof(int) + ptr->arr_max_size * sizeof(double);
  int flags = IPC_CREAT | IPC_EXCL | FIX2LONG(permissions);

  if (-1 == (ptr->shmid = shmget( key, byte_size, flags))) {
    if (errno == EEXIST)
      ptr->shmid = shmget(key, byte_size, flags &= ~IPC_EXCL);
  }
```